File: Automatic-Sparse-Estimation/Automatic_sparse_estimation.py
import numpy as np

def ECDM(X1, X2):
    """
    Compute the threshold (Tn) for sparse estimation via ECDM methodology.
    
    Parameters
    ----------
    X1 : ndarray
        p1 x n matrix (class 1)
    X2 : ndarray
        p2 x n matrix (class 2)
    
    Returns
    -------
    Tn : float
        Sparsification threshold for cross-covariance matrix
    """
    if len(X1.shape) == 1:
        n = X1.shape[0]
        X1 = X1.reshape(1,n)
    else:
        n = X1.shape[1]
    if len(X2.shape) == 1:
        X2 = X2.reshape(1,n)

    n1 = int(np.ceil(n/2))
    n2 = n - n1
    u = 2 * n1 * n2 / ((n1 - 1) * (n2 - 1) * n * (n - 1))
    W = 0.0

    def V1(k, X):
        if np.floor(k/2) >= n1:
            index = np.arange(np.floor(k/2) - n1, np.floor(k/2)).astype(int)
        else:
            index = np.concatenate([np.arange(np.floor(k/2)), np.arange(np.floor(k/2) + n2, n)]).astype(int)
        return(X[:,index])
    def V2(k, X):
        if np.floor(k/2) <= n1:
            index = np.arange(np.floor(k/2), np.floor(k/2) + n2).astype(int)
        else:
            index = np.concatenate([np.arange(np.floor(k/2) - n1), np.arange(np.floor(k/2), n)]).astype(int)
        return(X[:,index])

    def H1_1(k):
        return(V1(k,X1).mean(axis = 1))
    def H2_1(k):
        return(V2(k,X1).mean(axis = 1))
    def H1_2(k):
        return(V1(k,X2).mean(axis = 1))
    def H2_2(k):
        return(V2(k,X2).mean(axis = 1))
    
    S = np.arange(3, 2 * n)
    M1 = [list(map(H1_1, S)), list(map(H1_2, S))]
    M2 = [list(map(H2_1, S)), list(map(H2_2, S))]

    def q_corr(i, j):
        q1_component = np.dot(X1[:,i - 1] - M1[0][i + j - 3], X1[:,j - 1] - M2[0][i + j - 3])
        q2_component = np.dot(X2[:,i - 1] - M1[1][i + j - 3], X2[:,j - 1] - M2[1][i + j - 3])
        return q1_component * q2_component
    
    for i in range(1, n):
        for j in range(i + 1, n + 1):
            W += q_corr(i, j)
    Tn = W * u

    return Tn

def sparse_cross_cov(X1, X2):
    """
    Sparse estimation of the high-dimensional cross-covariance matrix.

    Parameters
    ----------
    X1 : ndarray
        p1 x n matrix (class 1)
    X2 : ndarray
        p2 x n matrix (class 2)

    Returns
    -------
    sparse_cross_cov : ndarray
        Sparse cross-covariance matrix (p1 x p2)
    sample_cross_cov : ndarray
        Sample cross-covariance matrix (p1 x p2)
    Delta : float
        Sparsification threshold from ECDM
    """
    p1, n = X1.shape
    p2 = X2.shape[0]
    p = p1 * p2
    Delta = ECDM(X1=X1, X2=X2)

    X1 = (X1.T - X1.mean(axis=1)).T
    X2 = (X2.T - X2.mean(axis=1)).T
    sample_cross_cov = X1 @ X2.T / (n - 1)
    s_ast = sample_cross_cov.reshape(-1, order='F')
    sparse_cross_cov = np.zeros(p)
    ord = np.argsort(np.abs(s_ast))[::-1]
    cri = 0
    for r in range(p):
        cri += s_ast[ord[r]]**2
        sparse_cross_cov[ord[r]] = s_ast[ord[r]]
        if cri >= Delta:
            break
    sparse_cross_cov = sparse_cross_cov.reshape(p1, p2, order='F')
    return sparse_cross_cov, sample_cross_cov, Delta

def sparse_mean(X1, X2=None):
    """
    Sparse mean vector estimation (one- or two-class).

    Parameters
    ----------
    X1 : ndarray
        p x n1 matrix
    X2 : ndarray or None
        Optional p x n2 matrix (class 2). If None, estimates sparse mean of X1; otherwise
        estimates sparse difference between class means.

    Returns
    -------
    sparse_mean : ndarray
        Sparse mean vector (or sparse mean difference) (p,)
    sample_mean : ndarray
        Sample mean vector (or mean difference) (p,)
    Delta : float
        Sparsification threshold
    """
    p = X1.shape[0]
    n1 = X1.shape[1]

    if X2 is None:
        sample_mean = X1.mean(axis=1)
        trS1 = np.sum((X1.T - sample_mean) ** 2) / (n1 - 1)
        Delta = np.sum(sample_mean ** 2) - trS1 / n1
    else:
        n2 = X2.shape[1]
        mean1 = X1.mean(axis=1)
        mean2 = X2.mean(axis=1)
        sample_mean = mean1 - mean2
        trS1 = np.sum((X1.T - mean1) ** 2) / (n1 - 1)
        trS2 = np.sum((X2.T - mean2) ** 2) / (n2 - 1)
        Delta = np.sum(sample_mean ** 2) - trS1 / n1 - trS2 / n2

    sort_idx = np.argsort(np.abs(sample_mean))[::-1]
    sparse_mean = np.zeros(p)
    cumulative_energy = 0.0

    for idx in range(p):
        element_idx = sort_idx[idx]
        cumulative_energy += sample_mean[element_idx] ** 2
        sparse_mean[element_idx] = sample_mean[element_idx]
        if cumulative_energy >= Delta:
            break

    return sparse_mean, sample_mean, Delta

File: Automatic-Sparse-Estimation/Automatic_sparse_estimation.r
#' Compute the threshold (Tn) for sparse estimation via ECDM methodology.
#'
#' @param X1 p1 x n matrix (class 1)
#' @param X2 p2 x n matrix (class 2)
#'
#' @return Tn: float
#'   Sparsification threshold for cross-covariance matrix
#'
#' @keywords internal
ECDM <- function(X1, X2) {
  if (is.null(dim(X1))) {
    X1 <- t(as.matrix(X1))
  }
  if (is.null(dim(X2))) {
    X2 <- t(as.matrix(X2))
  }
  
  p <- list(dim(X1)[1], dim(X2)[1])
  n <- dim(X1)[2]
  n1 <- as.integer(ceiling(n / 2))
  n2 <- n - n1
  
  K <- c(3:(2 * n - 1))
  L <- length(K)
  Y1 <- array(0, dim = c(2, L, p[[1]]))
  Y2 <- array(0, dim = c(2, L, p[[2]]))
  
  for (l in 1:L) {
    V <- list()
    dv <- as.integer(floor(K[l] / 2))
    
    if (dv >= n1) {
      id <- c((dv - n1 + 1):dv)
      V <- append(V, list(id))
    } else {
      id <- append(c(1:dv), c((dv + n2 + 1):n))
      V <- append(V, list(id))
    }
    
    if (dv <= n1) {
      id <- c((dv + 1):(dv + n2))
      V <- append(V, list(id))
    } else {
      id <- append(c(1:(dv - n1)), c((dv + 1):n))
      V <- append(V, list(id))
    }
    
    for (i in 1:2) {
      if (is.null(dim(X1[, V[[i]]]))) {
        Y1[i, l, ] <- apply(t(as.matrix(X1[, V[[i]]])), 1, mean)
      } else {
        Y1[i, l, ] <- apply(X1[, V[[i]]], 1, mean)
      }
      if (is.null(dim(X2[, V[[i]]]))) {
        Y2[i, l, ] <- apply(t(as.matrix(X2[, V[[i]]])), 1, mean)
      } else {
        Y2[i, l, ] <- apply(X2[, V[[i]]], 1, mean)
      }
    }
  }
  
  u <- n1 * n2 / ((n1 - 1) * (n2 - 1))
  t <- 0
  for (j in 1:n) {
    for (i in 1:j) {
      if (i != j) {
        t <- t + as.numeric((X1[, i] - Y1[1, (i + j - 2), ]) %*% (X1[, j] - Y1[2, (i + j - 2), ])) * 
                 as.numeric((X2[, i] - Y2[1, (i + j - 2), ]) %*% (X2[, j] - Y2[2, (i + j - 2), ]))
      }
    }
  }
  Tn <- 2 * u / (n * (n - 1)) * t
  
  Tn
}

#' Sparse estimation of the high-dimensional cross-covariance matrix.
#'
#' @param X1 p1 x n matrix (class 1)
#' @param X2 p2 x n matrix (class 2)
#'
#' @return A list containing:
#'   \itemize{
#'     \item sparse_cross_cov: Sparse cross-covariance matrix (p1 x p2)
#'     \item sample_cross_cov: Sample cross-covariance matrix (p1 x p2)
#'     \item Delta: Sparsification threshold from ECDM
#'   }
#'
#' @export
sparse_cross_cov <- function(X1, X2) {
  p1 <- nrow(X1)
  n <- ncol(X1)
  p2 <- nrow(X2)
  p <- p1 * p2

  Delta <- ECDM(X1, X2)

  X1c <- sweep(X1, 1, rowMeans(X1))
  X2c <- sweep(X2, 1, rowMeans(X2))

  sample_cross_cov <- (X1c %*% t(X2c)) / (n - 1)
  # Convert to column-major order (same as Python's order='F')
  cross_cov_vec <- as.vector(sample_cross_cov)

  sort_idx <- order(abs(cross_cov_vec), decreasing = TRUE)
  sparse_cross_cov <- rep(0, p)
  cri <- 0

  for (idx in seq_len(p)) {
    element_idx <- sort_idx[idx]
    cri <- cri + cross_cov_vec[element_idx]^2
    sparse_cross_cov[element_idx] <- cross_cov_vec[element_idx]
    if (cri >= Delta) break
  }

  # Reconstruct matrix in column-major order (byrow=FALSE is default)
  sparse_cross_cov_mat <- matrix(sparse_cross_cov, nrow = p1, ncol = p2, byrow = FALSE)
  list(sparse_cross_cov = sparse_cross_cov_mat, sample_cross_cov = sample_cross_cov, Delta = Delta)
}

#' Sparse mean vector estimation (one- or two-class).
#'
#' @param X1 p x n1 matrix
#' @param X2 optional p x n2 matrix (class 2). If missing, estimates sparse mean of X1;
#'           if provided, estimates sparse difference between class means.
#'
#' @return A list containing:
#'   \itemize{
#'     \item sparse_mean: Sparse mean vector (or sparse mean difference) (p,)
#'     \item sample_mean: Sample mean vector (or mean difference) (p,)
#'     \item Delta: Sparsification threshold
#'   }
#'
#' @export
sparse_mean <- function(X1, X2 = NULL) {
  p <- nrow(X1)
  n1 <- ncol(X1)

  if (is.null(X2)) {
    sample_mean <- rowMeans(X1)
    trS1 <- sum((t(X1) - rep(sample_mean, each = n1))^2) / (n1 - 1)
    Delta <- sum(sample_mean^2) - trS1 / n1
  } else {
    n2 <- ncol(X2)
    mean1 <- rowMeans(X1)
    mean2 <- rowMeans(X2)
    sample_mean <- mean1 - mean2
    trS1 <- sum((t(X1) - rep(mean1, each = n1))^2) / (n1 - 1)
    trS2 <- sum((t(X2) - rep(mean2, each = n2))^2) / (n2 - 1)
    Delta <- sum(sample_mean^2) - trS1 / n1 - trS2 / n2
  }

  sort_idx <- order(abs(sample_mean), decreasing = TRUE)
  sparse_mean <- rep(0, p)
  cri <- 0

  for (idx in seq_len(p)) {
    element_idx <- sort_idx[idx]
    cri <- cri + sample_mean[element_idx]^2
    sparse_mean[element_idx] <- sample_mean[element_idx]
    if (cri >= Delta) break
  }

  list(sparse_mean = sparse_mean, sample_mean = sample_mean, Delta = Delta)
}
